src/aspire/source/eman.py

```diff
@@ -0,0 +1,255 @@
+                    # if self.centers, subtract off half of the particle size to get the lower-left corner position
+                    # otherwise, we are reslicing from old_size to new_size, and we add half the difference of the two sizes to the lower left coordinates
+                    trim_length = (
+                        -(new_size // 2) if self.centers else (old_size - new_size) // 2
+                    )
+                    coords[0] += trim_length
+                    coords[1] += trim_length
+                    coords[2] = coords[3] = new_size
+
+        if self.centers:
+            assert (
+                particle_size > 0
+            ), "When constructing an EmanSource with coordinates of par\
+ticle centers, a particle size must be specified."
+            # recompute coordinates to account for the fact that we were given centers
+            size_particles(self.mrc2coords, particle_size)
+            L = particle_size
+        else:
+            # open first coord file to get the particle size in the file
+            first_coord_filepath = coord_paths[0]
+            with open(first_coord_filepath, "r") as coord_file:
+                first_line = coord_file.readlines()[0]
+                L = int(first_line.split()[2])
+                other_side = int(first_line.split()[3])
+
+            # firstly, ensure square particles
+            if not L == other_side:
+                logger.error(
+                    "Particle size in coordinates file is {L}x{other_side}, but only square particle images are supported."
+                )
+                raise ValueError
+            # if particle_size specified by user, we will recompute the coordinates around the center of the particle
+            if not particle_size == 0:
+                logger.info(
+                    f"Overriding particle size of {L}x{L} specified in coordinates file."
+                )
+                size_particles(self.mrc2coords, particle_size, L)
+                L = particle_size
+
+        logger.info(f"Particle size = {L}x{L}")
+        self._original_resolution = L
+
+        original_n = sum([len(self.mrc2coords[x]) for x in self.mrc2coords])
+
+        # Lastly, exclude particle coordinate boxes that do not fit into the micrograph dimensions
+        for _mrc, coordsList in self.mrc2coords.items():
+            out_of_range = []
+            for i in range(len(coordsList)):
+                coord = coordsList[i]
+                start_x, start_y, size_x, size_y = (
+                    coord[0],
+                    coord[1],
+                    coord[2],
+                    coord[3],
+                )
+                if (
+                    start_x < 0
+                    or start_y < 0
+                    or (start_x + size_x >= self.X)
+                    or (start_y + size_y >= self.Y)
+                ):
+                    out_of_range.append(i)
+            # pop in reverse order to avoid messing up indices
```

Comment:
If I pop, e.g. index `0`, then the indices of `coordsList` get shifted back by one, so the `j` in `out_of_range` would no longer correspond to the indices of `coordsList` that need to be popped. If I go backwards though, then I know all the indices before the one I just popped aren't changed yet,

Reply:
but aren't all those after it changed?  Are you saying this is just for your internal indices used for no other purpose? (that would be fine I think)

Reply:
Yeah, these indices (`out_of_range`) are just internal to this iteration of the loop. So we delete the particles that don't fit from `coordsList`. Later, when we're accessing the particles by index in `images()`, this doesn't come up because the `mrc2coords` dict is flattened and we use the indices of that list. At this point the micrographs/coordinates aren't in a flattened state where global indices mean anything.

Reply:
The ones after it are changed but the `out_of_range` indices are sorted from largest to smallest, so it doesn't matter.